### DISCUSSION TITLE
High CPU usage due to clickhouse restarting? Memory too low?

### DISCUSSION THREAD

**Discussion by russellballestrini** [plausible/analytics#4740] [Self-Hosted Support]
Since for at least 14 days the clickhouse docker container has been using 100% CPU seemingly rebooting in a very tight loop. This makes a single core server utilize 4.0 load (4 times more than healthly). Plausible seems to function and collect my metrics but it seems like something is very broken.

performing a `docker logs -f` yields hundreds or thousands of lines per second, this seems related?

```
 (version 24.3.3.102 (official build))
2024.10.27 17:36:29.485178 [ 621 ] {} <Error> MergeTreeBackgroundExecutor: Exception while executing background task {322de74c-73d3-4c28-829c-71146847af75::20241025_22689_33730_313}: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 892.09 MiB (attempt to allocate chunk of 4361232 bytes), maximum: 861.81 MiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cbc851b
1. DB::Exception::Exception<char const*, char const*, String, long&, String, char const*, std::basic_string_view<char, std::char_traits<char>>>(int, FormatStringHelperImpl<std::type_identity<char const*>::type, std::type_identity<char const*>::type, std::type_identity<String>::type, std::type_identity<long&>::type, std::type_identity<String>::type, std::type_identity<char const*>::type, std::type_identity<std::basic_string_view<char, std::char_traits<char>>>::type>, char const*&&, char const*&&, String&&, long&, String&&, char const*&&, std::basic_string_view<char, std::char_traits<char>>&&) @ 0x000000000cbde72a
2. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbde368
3. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9
4. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9
5. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9
6. Allocator<false, false>::alloc(unsigned long, unsigned long) @ 0x000000000cb9b7ed
7. DB::Memory<Allocator<false, false>>::alloc(unsigned long) @ 0x000000000cc220a2
```

Please let me know what I should do. I grew my swap file from 1024 to 2048 but that didn't help.

**Comment by ruslandoga** [comment]
👋 @russellballestrini 

> Code: 241. DB::Exception: Memory limit (total) exceeded: would use 892.09 MiB

ClickHouse is running out of memory. The easiest solution is to give it more RAM.

**Reply by russellballestrini** [reply]
Cool is this something with the docker compose override file? Do you have a link to a hint?

**Reply by ruslandoga** [reply]
Unless you limited RAM for the ClickHouse container, the way only way I know of to increase RAM is to get a bigger instance on your hosting provider :)

**Reply by russellballestrini** [reply]
I added a bigger swapfile is this not utilized?

**Reply by ruslandoga** [reply]
It should be. But using swap would slow ClickHouse down. I remember enabling it on my 2GB RAM instance and things got somewhat similar to what you describe: constant high CPU load. In addition I also had a very slow shell.

**Reply by ruslandoga** [reply]
https://chistadata.com/clickhouse-linux-swaps/ might be helpful.

**Reply by russellballestrini** [reply]
Hmmm. Maybe. I seems like the CPU is being utilized because something is restarting in a tight loop trying to allocate memory. That leads me to believe the swap isn't being used or if it is, it is used after the exception is raised?

**Reply by russellballestrini** [reply]
I will double memory from 1G to 2G and see of this resolves, otherwise 4G. That said this is a rather light weight plausible  cluster, no more than a hundred web requests per minute at the max.

**Reply by ruslandoga** [reply]
Check `docker stats` or `htop`. Maybe the extra amount of swap you added is not enough, and 826MB is just the last allocation that fails.

**Reply by russellballestrini** [reply]
For testing (I should be able to revert) I upgraded from 1G to 2G, now no swap is being used. and load is 1.0 (down from 4.0 due to restarting over and over.).

```
systemd+    2199 31.5 25.1 13460708 506424 ?     Ssl  08:35   6:40 /usr/bin/clickhouse-server --config-file=/etc/clickhouse-server/config.xml
```

this service is bursty and consumes 30-70% cpu, on a single core system it ends up idling around 1.0 instead of closer to 0.1 ... Is this normal? I'm seeing much less logs now but still some.

```
2024.10.28 12:54:57.253963 [ 633 ] {} <Error> MergeTreeBackgroundExecutor: Exception while executing background task {322de74c-73d3-4c28-829c-71146847af75::20241028_52864_57099_302}: Code: 241. DB::
Exception: Memory limit (total) exceeded: would use 1.73 GiB (attempt to allocate chunk of 4361232 bytes), maximum: 1.73 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time o
r overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):  
```

like I said much less frequent errors now but they still happen. Did my clickhouse have some big object that it needed to process and now I can scale back down or will the issue come back?

**Reply by ruslandoga** [reply]
> Did my clickhouse have some big object that it needed to process 

Yes, the message seems to come from a background job that merges merge tree parts.

> now I can scale back down or will the issue come back?

It would probably happen again, once the big parts get selected to be merged again.

**Reply by ruslandoga** [reply]
https://clickhouse.com/docs/en/operations/tips#using-less-than-16gb-of-ram might be helpful.

Is there any stacktrace in the error message?

> Stack trace (when copying this message, always include the lines below):

It might help us identify which configuration options for which subsystem to tune.

---

And also setting these lower than defaults might help:
- https://clickhouse.com/docs/en/operations/settings/query-complexity#settings_max_memory_usage
- https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#merges_mutations_memory_usage_to_ram_ratio
- https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#merges_mutations_memory_usage_soft_limit
- https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#background_pool_size

**Reply by russellballestrini** [reply]
```
 (version 24.3.3.102 (official build))                                                                                                                                                       [56/9238]
2024.10.28 12:54:57.683968 [ 628 ] {} <Error> MergeTreeBackgroundExecutor: Exception while executing background task {322de74c-73d3-4c28-829c-71146847af75::20241028_52864_57099_302}: Code: 241. DB::
Exception: Memory limit (total) exceeded: would use 1.73 GiB (attempt to allocate chunk of 4361232 bytes), maximum: 1.73 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time o
r overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):                                                           
                                                                                                                                                                                                      
0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cbc851b                                                                                                           
1. DB::Exception::Exception<char const*, char const*, String, long&, String, char const*, std::basic_string_view<char, std::char_traits<char>>>(int, FormatStringHelperImpl<std::type_identity<char co
nst*>::type, std::type_identity<char const*>::type, std::type_identity<String>::type, std::type_identity<long&>::type, std::type_identity<String>::type, std::type_identity<char const*>::type, std::t
ype_identity<std::basic_string_view<char, std::char_traits<char>>>::type>, char const*&&, char const*&&, String&&, long&, String&&, char const*&&, std::basic_string_view<char, std::char_traits<char>
>&&) @ 0x000000000cbde72a                                                                                                                                                                             
2. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbde368                                                                                                                  
3. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9                                                                                                                  
4. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9                                                                                                                  
5. MemoryTracker::allocImpl(long, bool, MemoryTracker*, double) @ 0x000000000cbddda9                                                                                                                  
6. Allocator<false, false>::alloc(unsigned long, unsigned long) @ 0x000000000cb9b7ed                                                                                                                  
7. DB::Memory<Allocator<false, false>>::alloc(unsigned long) @ 0x000000000cc220a2                                                                                                                     
8. DB::WriteBufferFromFileDescriptor::WriteBufferFromFileDescriptor(int, unsigned long, char*, std::shared_ptr<DB::Throttler>, unsigned long, String) @ 0x000000000cc43634                            
9. DB::WriteBufferFromFile::WriteBufferFromFile(String const&, unsigned long, int, std::shared_ptr<DB::Throttler>, unsigned int, char*, unsigned long) @ 0x000000000cd37da4
10. DB::DiskLocal::writeFile(String const&, unsigned long, DB::WriteMode, DB::WriteSettings const&) @ 0x0000000010327773                                                                              
11. DB::DataPartStorageOnDiskFull::writeFile(String const&, unsigned long, DB::WriteMode, DB::WriteSettings const&) @ 0x0000000011c6c943                                                              
12. DB::MergeTreeDataPartWriterOnDisk::Stream<false>::Stream(String const&, std::shared_ptr<DB::IDataPartStorage> const&, String const&, String const&, String const&, String const&, std::shared_ptr<
DB::ICompressionCodec> const&, unsigned long, std::shared_ptr<DB::ICompressionCodec> const&, unsigned long, DB::WriteSettings const&) @ 0x0000000011e1491e
13. void std::__function::__policy_invoker<void (DB::ISerialization::SubstreamPath const&)>::__call_impl<std::__function::__default_alloc_func<DB::MergeTreeDataPartWriterWide::addStreams(DB::NameAnd
TypePair const&, std::shared_ptr<DB::IAST> const&)::$_0, void (DB::ISerialization::SubstreamPath const&)>>(std::__function::__policy_storage const*, DB::ISerialization::SubstreamPath const&) @ 0x000
0000011e2578d                                                                                      
14. DB::SerializationSparse::enumerateStreams(DB::ISerialization::EnumerateStreamsSettings&, std::function<void (DB::ISerialization::SubstreamPath const&)> const&, DB::ISerialization::SubstreamData 
const&) const @ 0x00000000101371af
15. DB::ISerialization::enumerateStreams(std::function<void (DB::ISerialization::SubstreamPath const&)> const&, std::shared_ptr<DB::IDataType const> const&, COW<DB::IColumn>::immutable_ptr<DB::IColu
mn> const&) const @ 0x00000000100f2c01
16. DB::MergeTreeDataPartWriterWide::MergeTreeDataPartWriterWide(std::shared_ptr<DB::IMergeTreeDataPart> const&, DB::NamesAndTypesList const&, std::shared_ptr<DB::StorageInMemoryMetadata const> cons
t&, std::vector<std::shared_ptr<DB::IMergeTreeIndex const>, std::allocator<std::shared_ptr<DB::IMergeTreeIndex const>>> const&, std::vector<std::shared_ptr<DB::IStatistic>, std::allocator<std::share
d_ptr<DB::IStatistic>>> const&, String const&, std::shared_ptr<DB::ICompressionCodec> const&, DB::MergeTreeWriterSettings const&, DB::MergeTreeIndexGranularity const&) @ 0x0000000011e1d0cb
17. DB::MergeTreeDataPartWide::getWriter(DB::NamesAndTypesList const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::vector<std::shared_ptr<DB::IMergeTreeIndex const>, std::allocat
or<std::shared_ptr<DB::IMergeTreeIndex const>>> const&, std::vector<std::shared_ptr<DB::IStatistic>, std::allocator<std::shared_ptr<DB::IStatistic>>> const&, std::shared_ptr<DB::ICompressionCodec> c
onst&, DB::MergeTreeWriterSettings const&, DB::MergeTreeIndexGranularity const&) @ 0x0000000011e073a4
18. DB::MergedBlockOutputStream::MergedBlockOutputStream(std::shared_ptr<DB::IMergeTreeDataPart> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::NamesAndTypesList const&, std:
:vector<std::shared_ptr<DB::IMergeTreeIndex const>, std::allocator<std::shared_ptr<DB::IMergeTreeIndex const>>> const&, std::vector<std::shared_ptr<DB::IStatistic>, std::allocator<std::shared_ptr<DB
::IStatistic>>> const&, std::shared_ptr<DB::ICompressionCodec>, std::shared_ptr<DB::MergeTreeTransaction> const&, bool, bool, DB::WriteSettings const&, DB::MergeTreeIndexGranularity const&) @ 0x0000
000011f626a4
19. DB::MergeTask::ExecuteAndFinalizeHorizontalPart::prepare() @ 0x0000000011d0d962
20. DB::MergeTask::ExecuteAndFinalizeHorizontalPart::execute() @ 0x0000000011d13980
21. DB::MergeTask::execute() @ 0x0000000011d17e79 
```

**Reply by russellballestrini** [reply]
I see many tunable linked. thanks. any recommendations & where do I config it?

**Reply by ruslandoga** [reply]
You can add XML configs as bind mounts like https://github.com/plausible/community-edition/blob/5eb6ea318cfc57fcce520af1e568f56dac805916/compose.yml#L19-L22 and ClickHouse would merge them into the final configuration.

**Comment by salomvary** [comment]
I am having the same issue, Clickhouse is using 100% repeatedly logs `MEMORY_LIMIT_EXCEEDED` errors. 

`Memory limit (total) exceeded: would use 3.36 GiB (attempt to allocate chunk of 4361184 bytes), current RSS 1.66 GiB, maximum: 3.35 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED)`

Full stack trace here: https://gist.github.com/salomvary/34c1cf60f56e52a5e14433115f226512

The amount of memory (~4GB) available is reasonable compared to how small the database is:

```
SELECT
    database,
    `table`,
    formatReadableSize(sum(data_compressed_bytes) AS size) AS compressed,
    formatReadableSize(sum(data_uncompressed_bytes) AS usize) AS uncompressed,
    round(usize / size, 2) AS compr_rate,
    sum(rows) AS rows,
    count() AS part_count
FROM system.parts
WHERE (active = 1) AND (database LIKE 'plausible_events_db') AND (`table` LIKE '%')
GROUP BY
    database,
    `table`
ORDER BY size DESC

Query id: 3274f20f-2e0e-4ad3-ab87-f99b05949c12

   ┌─database────────────┬─table───────────┬─compressed─┬─uncompressed─┬─compr_rate─┬───rows─┬─part_count─┐
1. │ plausible_events_db │ location_data   │ 2.89 MiB   │ 3.92 MiB     │       1.35 │ 203370 │          1 │
2. │ plausible_events_db │ sessions_v2     │ 4.77 KiB   │ 10.61 KiB    │       2.22 │     61 │          1 │
3. │ plausible_events_db │ events_v2       │ 4.25 KiB   │ 11.15 KiB    │       2.62 │     90 │          1 │
4. │ plausible_events_db │ ingest_counters │ 710.00 B   │ 1.85 KiB     │       2.67 │     77 │          1 │
   └─────────────────────┴─────────────────┴────────────┴──────────────┴────────────┴────────┴────────────┘

4 rows in set. Elapsed: 0.015 sec. 
```

Using `clickhouse/clickhouse-server:24.10.2.80` Docker image.

Also getting `get_mempolicy: Operation not permitted` to stderr, which seems to be this bug: https://github.com/ClickHouse/ClickHouse/issues/70960. Not sure if relevant here.

**Reply by ruslandoga** [reply]
👋 @salomvary 

Maybe there are some problems with the recent ClickHouse releases or they just got more memory hungry. Could you please try downgrading to see if it gets any better?

**Reply by ruslandoga** [reply]
To something like [clickhouse/clickhouse-server:24.3.3.102-alpine](https://github.com/plausible/community-edition/blob/v2.1.4/compose.yml#L14) or maybe even [clickhouse/clickhouse-server:23.3.7.5-alpine](https://github.com/plausible/community-edition/blob/v2.0.0/docker-compose.yml#L17). I don't think people reported any memory usage problems with the latter one at all.

**Reply by salomvary** [reply]
@ruslandoga Downgrading to `clickhouse/clickhouse-server:24.3.3.102` indeed helped, I no longer get any errors logged and the CPU usage of `/usr/bin/clickhouse-server` is down to ~15%. Thanks a ton for the tip! Shall we report this as a bug somewhere?

**Reply by ruslandoga** [reply]
> Shall we report this as a bug somewhere?

That would be a nice thing to do. But first we'd need to check if anything similar has been reported already and they would probably want a reproducible example (that doesn't involve installing Plausible). Maybe "cloning" the container volume and sharing it would work. We'd need to make sure that the problem still appears on that minimal setup.

**Reply by salomvary** [reply]
I already see whatever ticket I file for Clickhouse getting closed as wontfix because:

> If your system has less than 16 GB of RAM, you may experience various memory exceptions because default settings do not match this amount of memory. 

https://clickhouse.com/docs/en/operations/tips#using-less-than-16gb-of-ram

Sounds like this should have never worked without extra tweaking of Clickhouse settings?

**Reply by ruslandoga** [reply]
But it did :) I was running Plausible on a free-tier AWS EC2 instance until earlier this year when it started to get OOMed (after upgrading from ClickHouse 23.3.7.5-alpine), then I switched to a Hetzner instance. I have an issue open to try and find ways to decrease memory usage in CE by default: https://github.com/plausible/community-edition/issues/185 -- it would be great if the community contributes their ideas / experience :)

**Reply by salomvary** [reply]
Update: I am still getting `MEMORY_LIMIT_EXCEEDED` errors on 24.3.3.102, but not continuously, more like every few minutes.

**Reply by charleslcso** [reply]
With the help of @ruslandoga I got my first instance of Plausible working last night. Immediately I noticed CPU usage is very high. Nothing hitting Plausible, but CPU usage goes to almost 50%. I thought it might be doing some background stuff, and let it sat there.

12hrs later the CPU usage hasn't dropped. 

Found out Postfix is triggering Clickhouse.

Captured from 'top':
16247 postfix   20   0   12.1g 540212 154768 S  31.6  25.8 356:19.84 clickhouse-serv

**Reply by ruslandoga** [reply]
👋 @charleslcso 

1. We don't use Postfix anywhere
2. `postfix` in your `top` line is the name of the user running `clickhouse-serv` and not the name of the program

We have an issue open to try and lower the default resource requirements for ClickHouse https://github.com/plausible/community-edition/issues/185 but unless you are experiencing OOMs or other problems, having ClickHouse consume available resources is not an issue. After all, you have already paid for these resources (assuming you are using a cloud provider), so you might as well give them to ClickHouse :)

**Reply by charleslcso** [reply]
Yes, point 2 is what's been observed. I thought Plausible would use postfix to run clickhouse-serv...

Weird. I never heard of clickhouse before using Plausible, so I never configured it.

Letting Clickhouse use resources is OK with me. But user Postfix using Clickhouse? I believe I need to get to bottom of this...

**Reply by ruslandoga** [reply]
> But user Postfix using Clickhouse? I believe I need to get to bottom of this...

In the default [setup](https://github.com/plausible/community-edition/blob/5eb6ea318cfc57fcce520af1e568f56dac805916/compose.yml#L13-L14) ClickHouse is being run in a container, and the ClickHouse container uses UID 101 by [default,](https://github.com/ClickHouse/ClickHouse/blob/617af1d3c07cf598b4d7ae014c5122bf92e5c4a5/docker/server/Dockerfile.alpine#L48-L49) which might correspond to an existing user on the host. You can configure different UID / GID with `compose.override.yml` using `user:` directive.
